pkg/syncer/index.go

```diff
@@ -0,0 +1,195 @@
+	toposort "github.com/philopon/go-toposort"
+	"k8s.io/klog"
+
+	"github.com/bitnami-labs/charts-syncer/pkg/chart"
+	"github.com/bitnami-labs/charts-syncer/pkg/utils"
+)
+
+// Chart describes a chart, including dependencies
+type Chart struct {
+	Name         string
+	Version      string
+	Dependencies []string
+
+	TgzPath string
+}
+
+// ChartIndex is a map linking a chart reference with its Chart
+type ChartIndex map[string]*Chart
+
+// getIndex returns the chart index
+func (s *Syncer) getIndex() ChartIndex {
+	if s.index == nil {
+		s.index = make(ChartIndex)
+	}
+	return s.index
+}
+
+// Add adds a chart in the index
+func (i ChartIndex) Add(id string, chart *Chart) error {
+	if _, ok := i[id]; ok {
+		return errors.Errorf("%q is already indexed", id)
+	}
+	i[id] = chart
+	return nil
+}
+
+// Get returns an index chart
+func (i ChartIndex) Get(id string) *Chart {
+	if c, ok := i[id]; ok {
+		return c
+	}
+	return nil
+}
+
+// loadChartsIndex loads the charts map into the index from the source repo
+func (s *Syncer) loadChartsIndex(charts ...string) error {
+	if len(charts) == 0 {
+		if !s.autoDiscovery {
+			return errors.Errorf("unable to discover charts to sync")
+		}
+
+		srcCharts, err := s.cli.src.List()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		charts = srcCharts
+	}
+	// Sort chart names
+	sort.Strings(charts)
+
+	// Create basic layout for date and parse flag to time type
+	publishingThreshold, err := utils.GetDateThreshold(s.fromDate)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Iterate over charts in source index
+	var errs error
+	for _, name := range charts {
+		versions, err := s.cli.src.ListChartVersions(name)
+		if err != nil {
+			errs = multierror.Append(errs, errors.Trace(err))
+			continue
+		}
+
+		klog.V(3).Infof("Indexing %q charts...", name)
+		for _, version := range versions {
+			details, err := s.cli.src.GetChartDetails(name, version)
+			if err != nil {
+				errs = multierror.Append(errs, errors.Trace(err))
+				continue
+			}
+
+			id := fmt.Sprintf("%s-%s", name, version)
+			if details.PublishedAt.Before(publishingThreshold) {
+				klog.V(4).Infof("Skipping %q chart: Published before %q", id, publishingThreshold.String())
+				continue
+			}
+
+			if ok, err := s.cli.dst.Has(name, version); err != nil {
+				errs = multierror.Append(errs, errors.Trace(err))
+				continue
+			} else if ok {
+				klog.V(4).Infof("Skipping %q chart: Already synced", id)
+				continue
+			}
+
+			if ch := s.getIndex().Get(id); ch != nil {
+				klog.V(4).Infof("Skipping %q chart: Already indexed", id)
+				continue
+			}
+
+			if err := s.loadChart(name, version); err != nil {
+				errs = multierror.Append(errs, errors.Trace(err))
+				continue
+			}
+		}
+	}
+
+	return errors.Trace(errs)
+}
+
+// loadChart loads a chart in the chart index map
+func (s *Syncer) loadChart(name string, version string) error {
+	id := fmt.Sprintf("%s-%s", name, version)
+	if ch := s.getIndex().Get(id); ch != nil {
```

Comment:
aren't you checking this already before calling the function? 

Line 109 of this same file

Reply:
yes! The difference is that `loadChart` is a recursive function and it will be invoked again for each dependency. It makes sense that different "tier1" charts use the same "tier2" chart dependencies. This check will make the method to skip already indexed charts.

Should I clarify it with a comment?

Reply:
yes, please!